### DISCUSSION TITLE
Breakouts for MAIANA

### DISCUSSION THREAD

**Discussion by peterantypas** [peterantypas/maiana#53] [Ideas]
I've been thinking about all the possible interfaces that MAIANA&trade; will need and it's plain as day (to me at least) that it's time to adopt the Adapter Design Pattern. This will pertain to the current (wired) version of the transponder.

Here is what I'm thinking:

![IMG_1023(1)](https://user-images.githubusercontent.com/1565933/110571060-d26e4f00-810b-11eb-9959-db3309b191fa.jpg)

The top board is standard. It has the RJ45 connector, an on/off and a "silent mode" switch, 12V power input (think Phoenix type screw terminals) and a USB->serial interface. If the device on the other end is just a computer, that's all one needs.

For all other interfaces, we have standard form factor for a plurality of custom adapter boards, each of which can implement one or more interfaces:

- NMEA0183 (RS422)
- NMEA2000
- WiFi / BLE
- RS232
- etc, etc

So, rather than provide a single breakout box with a nice but inflexible enclosure, I'd rather provide an expandable breakout board and let every boater mount it and wire it as they see fit.

I think this is where the Community can contribute without huge time and expense commitment. Maybe I can try to create a starting KiCAD project for the adapter boards, and let folks design whatever they want ...

Let's discuss.

**Comment by muebau** [comment]
I love this design. This project gets better every day! I would love if all the other projects for "boating" would be just half as smart as this one.

**Comment by revolteur** [comment]
I am liking the modular extension board concept and the possibilities it opens up!

Thinking through this concept, the list of adapters and their requirements, I can see the possibility of reducing them to one COM adapter and “Etc” adapter(s). Each adapter will support a 2 b2b connectors on opposite sides for expandability purposes.

COM Interface adapter: Considering the requirements of a NEMA2K adapter (uC, CAM), a NEMA183 adapter and WiFi adapter (ESP32), it might be just feasible to pull it off with 1 adapter instead of 3. The ESP32 supports CAN and Wi-Fi in once SoC. Timo Lappalainen N2K library has an implementation of the ESP32 CAN driver and NEMA2K and 183. 4 port phoenix connectors for wired NMEA. Wi-Fi is turned off by default and can be configured via the captive portal. This will cover most, if not all, Common boat interface setups that I can think of and have been discussing.

Etc adapters: For additional capabilities and un-common use cases.

The new functionality/concept of the breakout board concept merits the 2 unit design in this case.
Hoping you are starting to like the "standard RJ45 board" a bit more :)
I am liking the path this is going.

**Comment by marcelrv** [comment]
I like the idea, for having an expansion board. 
Having said that, my experience with connectors on my boat is very poor, as they all get corroded over the years. 
Hence for the most common usecase I still would be great to have them on the standard breakout board instead of the expansion board.
as suggested by @revolteur an ESP32 + standard serial (e.g. max3232) would already account for 99% of the usecases: (5v serial, NEMA0183 serial, wifi,BLE, NEMA2K)

**Comment by peterantypas** [comment]
Alright, the breakout is in progress. Ditched the USB interface from it to keep it simple:

<img width="304" alt="Screen Shot 2021-03-13 at 12 42 56 PM" src="https://user-images.githubusercontent.com/1565933/111043841-b4e60180-83f9-11eb-92eb-55db032385d9.png">

It will supply 12V, 5V and 3.3V to the expansion board (assuming it's powered on via the switch). The 5V and 3.3V LDOs can handle 500mA each.

**Comment by revolteur** [comment]
As an alternative to the JeeLabs (ESP32 Serial Link) library, the IotWebConf Library might be worth considering. It provides a non-blocking standalone WiFi/AP web configuration portal among other relevant features not available in the WifiManager library. It's being used in a similar use case by the open satellite network TinyGS to configure ESP32/LORA base station receivers. Something to consider: https://github.com/prampec/IotWebConf
